bentoml/marshal/marshal.py

```diff
@@ -213,23 +216,36 @@ def fetch_sema(self):
             self._outbound_sema = NonBlockSema(self.outbound_workers)
         return self._outbound_sema
 
-    def get_client(self):
-        from aiohttp import ClientSession, DummyCookieJar, UnixConnector, TCPConnector
+    def get_conn(self) -> "BaseConnector":
+        import aiohttp
 
-        if self._client is None:
-            jar = DummyCookieJar()
+        if self._conn is None or self._conn.closed:
             if self.outbound_unix_socket:
-                conn = UnixConnector(path=self.outbound_unix_socket,)
+                self._conn = aiohttp.UnixConnector(path=self.outbound_unix_socket,)
             else:
-                conn = TCPConnector(limit=30)
-            self._client = ClientSession(
-                connector=conn, auto_decompress=False, cookie_jar=jar,
+                self._conn = aiohttp.TCPConnector(limit=30)
+        return self._conn
+
+    def get_client(self):
+        import aiohttp
+
+        if self._client is None or self._client.closed:
+            jar = aiohttp.DummyCookieJar()
+            self._client = aiohttp.ClientSession(
+                connector=self.get_conn(),
+                auto_decompress=False,
+                cookie_jar=jar,
+                connector_owner=False,
             )
         return self._client
 
     def __del__(self):
+        loop = asyncio.get_event_loop()
         if getattr(self, '_client', None) is not None and not self._client.closed:
-            self._client.close()
+            loop.create_task(self._client.close())
+
+        # https://docs.aiohttp.org/en/stable/client_advanced.html#graceful-shutdown
+        loop.create_task(asyncio.sleep(0))
```

Comment:
what's the difference between this and what was suggested in the docs: `loop.run_until_complete(asyncio.sleep(0))`?

Reply:
When the loop is already running, we append a new task to it rather than `loop.run_until_complete`

Reply:
> Note that the appropriate amount of time to wait will vary from application to application.

It seems like no amount of time is reliable. Can we make the sleep time configurable and default to something conservative? Also, `aiohttp` 4.0 has resolved this issue. Let's keep it in the comments so when we upgrade we will know to fix here as well.

Reply:
@ssheng Ends up by hooking on the aiohttp `on_cleanup` to solve this.